tsdb/chunkenc/float_histogram.go

```diff
@@ -44,6 +44,29 @@ func NewFloatHistogramChunk() *FloatHistogramChunk {
 	return &FloatHistogramChunk{b: bstream{stream: b, count: 0}}
 }
 
+// NewPrepopulatedFloatHistogramChunk returns the decoded version of a previously-encoded chunk.
+func NewPrepopulatedFloatHistogramChunk(b []byte) (*FloatHistogramChunk, error) {
+	chk := &FloatHistogramChunk{b: bstream{stream: b}}
+	it, ok := chk.Iterator(nil).(*floatHistogramIterator)
+	if !ok {
+		panic("unexpected iterator type") // This should never happen.
+	}
+
+	for valType := it.Next(); valType != ValNone; valType = it.Next() {
+	}
+	if err := it.Err(); err != nil {
+		return nil, err
+	}
+
+	if it.br.valid%8 == 0 && (it.br.valid != 0 || it.br.streamOffset < len(it.br.stream)) {
+		chk.b.count = 8
+	} else {
+		chk.b.count = it.br.valid % 8
+	}
```

Comment:
So we are doing essentially the same in the (Float)HistogramChunk.Appender method, just without setting the bstream to the right state once we have iterated through the chunk.

Couldn't we just move that bstream handling into the Appender method? Then we wouldn't need those special `NewPrepopulated...` functions at all.

And then, of course, we could consider it for float chunks, too.

On the other hand, this appears now too obvious. Maybe there is another reason to enforce cutting a new chunk after snapshot loading? (Which would then mean we should also enforce cutting new chunks for histogram chunks… maybe that's worth paging @codesome …)

Reply:
I think for now, I'll remove the `NewPrepopulated...` in favor of doing the bstream count setting in the Appender methods while we wait for input from @codesome / @marctc / @jesusvazquez.

Reply:
Hi, @marctc and I are looking into this. I think the initial idea with the loadsnapshot from disks was to leave the head in the same state as it was prior to the snapshotting but its true that current behavior of initializing the appenders and iterating through all the samples to then just cut a new sample looks inefficient. 

@leizor's latest commit with the conservative approach looks good and perhaps thats enough for the scope of this PR. 

I'll continue to dig up to see if there was a reason for doing what we're doing at the moment, if not maybe we can just map the chunk but always start fresh after a snapshot which seems more efficient to me for now. In any case this could be a change for a follow up PR.

I should have a comment tomorrow about this.

Reply:
This got merged already but I did have a deeper look at it. I think we are good with the latest changes :+1: 

Will clarify with @codesome if we can be smarter about initializing the appenders after loading a snapshot onto memory cc @beorn7